LFS-821: Add default value option to computed fields

acrowthe opened the pull request:
- Default values can be specified in computed field expressions using the following syntax:
  `@{<question>:-<default value>}`
- Ommiting the `:-<default value>` option will cause the field to behave as before.
- Adding a default value will no longer halt execution if the specified question has no answer.
  Instead, the default value will be used
  
  Example: `return @{question1:-0}` will return question1's answer, or 0 if it has not been answered

acrowthe committed:
LFS-821: Add default value option to computed fields
- Default values can be specified in computed field expressions using the following syntax:
  `@{<question>|<default value>}`
- Ommiting the `|<default value>` option will cause the field to behave as before.
- Adding a default value will no longer halt execution if the specified question has no answer.
  Instead, the default value will be used

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -92,17 +97,26 @@ let ComputedQuestion = (props) => {
+      let defaultValue = null;
+      if (optionStart > -1 && optionStart < end) {
+        inputName = expr.substring(start + 2, optionStart);
+        defaultValue = expr.substring(optionStart + 2, end);
```
```suggestion
        defaultValue = expr.substring(optionStart + 1, end);
```

acrowthe replied:
Switch to optionStart.length

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -92,17 +97,26 @@ let ComputedQuestion = (props) => {
+        inputValues.push(getQuestionValue(inputName, form, defaultValue));
       }
-      // Remove the start and end tags from the expression
+      // Remove the start, end and default tagsfrom the expression
```
```suggestion
      // Remove the start, end and default tags from the expression
```

sdumitriu reviewed (changes requested):
I wonder if using bash syntax would be better (`${variable:-defaultValue}`. On one hand, it would allow adding support for other bash-like processing in the future, but on the other hand this isn't bash, and all the extra processing can be done in the eval expression via javascript. What do you think?

acrowthe commented:
Having bash-like syntax sounds like a good idea to me. I'm not sure we want to go for the full `${variable}` syntax rather than `@{}` since `${}` collides with javascript string templates, but `:-` rather than `|` makes sense

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -92,17 +97,26 @@ let ComputedQuestion = (props) => {
+      let inputName;
+      let defaultValue = null;
+      if (optionStart > -1 && optionStart < end) {
+        inputName = expr.substring(start + 2, optionStart);
```
```suggestion
        inputName = expr.substring(start + startTag.length, optionStart);
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -92,17 +97,26 @@ let ComputedQuestion = (props) => {
+      let defaultValue = null;
+      if (optionStart > -1 && optionStart < end) {
+        inputName = expr.substring(start + 2, optionStart);
+        defaultValue = expr.substring(optionStart + 2, end);
```
```suggestion
        defaultValue = expr.substring(optionStart + defaultTag.length, end);
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -92,17 +97,26 @@ let ComputedQuestion = (props) => {
+        inputName = expr.substring(start + 2, optionStart);
+        defaultValue = expr.substring(optionStart + 2, end);
+      } else {
+        inputName = expr.substring(start + 2, end);
```
```suggestion
        inputName = expr.substring(start + startTag.length, end);
```

acrowthe committed:
LFS-821: Add default value option to computed fields
Improve default option parsing
- Switch to bash-like syntax for defining options (`:-` instead of `|`)
- Improve code that removes start, end and option tags from computed expressions
  - Strip option tags and options
  - Ensure all characters in the expression are traversed
